plugins/inputs/fritzbox/fritzbox.go

```diff
@@ -0,0 +1,420 @@
+	Collect []string        `toml:"collect"`
+	Timeout config.Duration `toml:"timeout"`
+	tls.ClientConfig
+	Log             telegraf.Logger `toml:"-"`
+	deviceClients   []*tr064.Client
+	serviceHandlers map[string]serviceHandlerFunc
+}
+
+func defaultFritzbox() *Fritzbox {
+	return &Fritzbox{
+		URLs:            make([]string, 0),
+		Collect:         []string{"device", "wan", "ppp", "dsl", "wlan"},
+		Timeout:         defaultTimeout,
+		deviceClients:   make([]*tr064.Client, 0),
+		serviceHandlers: make(map[string]serviceHandlerFunc),
+	}
+}
+
+func (*Fritzbox) SampleConfig() string {
+	return sampleConfig
+}
+
+func (plugin *Fritzbox) Init() error {
+	err := plugin.initDeviceClients()
+	if err != nil {
+		return err
+	}
+	plugin.initServiceHandlers()
+	return nil
+}
+
+func (plugin *Fritzbox) initDeviceClients() error {
+	for _, rawUrl := range plugin.URLs {
+		parsedUrl, err := url.Parse(rawUrl)
+		if err != nil {
+			return err
+		}
+		client := tr064.NewClient(parsedUrl)
+		client.Debug = plugin.Log.Level().Includes(telegraf.Debug)
+		client.Timeout = time.Duration(plugin.Timeout)
+		tlsConfig, err := plugin.TLSConfig()
+		if err != nil {
+			return err
+		}
+		client.TlsConfig = tlsConfig
+		plugin.deviceClients = append(plugin.deviceClients, client)
+	}
+	return nil
+}
+
+func (plugin *Fritzbox) initServiceHandlers() {
+	if choice.Contains("device", plugin.Collect) {
+		plugin.serviceHandlers[deviceinfo.ServiceShortType] = plugin.gatherDeviceInfo
+	}
+	if choice.Contains("wan", plugin.Collect) {
+		plugin.serviceHandlers[wancommonifconfig.ServiceShortType] = plugin.gatherWanInfo
+	}
+	if choice.Contains("ppp", plugin.Collect) {
+		plugin.serviceHandlers[wanpppconn.ServiceShortType] = plugin.gatherPppInfo
+	}
+	if choice.Contains("dsl", plugin.Collect) {
+		plugin.serviceHandlers[wandslifconfig.ServiceShortType] = plugin.gatherDslInfo
+	}
+	if choice.Contains("wlan", plugin.Collect) {
+		plugin.serviceHandlers[wlanconfig.ServiceShortType] = plugin.gatherWlanInfo
+	}
+	if choice.Contains("hosts", plugin.Collect) {
+		plugin.serviceHandlers[hosts.ServiceShortType] = plugin.gatherHostsInfo
+	}
+}
+
+func (plugin *Fritzbox) Gather(acc telegraf.Accumulator) error {
+	var waitComplete sync.WaitGroup
+	for _, deviceClient := range plugin.deviceClients {
+		waitComplete.Add(1)
+		go func() {
+			defer waitComplete.Done()
+			plugin.gatherDevice(acc, deviceClient)
+		}()
+	}
+	waitComplete.Wait()
+	return nil
+}
+
+func (plugin *Fritzbox) gatherDevice(acc telegraf.Accumulator, deviceClient *tr064.Client) {
+	plugin.Log.Debugf("Querying %s", deviceClient.DeviceUrl.Redacted())
+	services, err := deviceClient.Services(tr064.DefaultServiceSpec)
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+	for _, service := range services {
+		serviceHandler := plugin.serviceHandlers[service.ShortType()]
+		if serviceHandler == nil {
+			continue
+		}
+		err := serviceHandler(acc, deviceClient, service)
+		if err != nil {
+			acc.AddError(err)
+			return
+		}
```

Comment:
The accumulator already performs this check
```suggestion
		acc.AddError(serviceHandler(acc, deviceClient, service))
```